filestore.go

```diff
@@ -0,0 +1,290 @@
+}
+
+// File represents a TileDB filestore file.
+type File struct {
+	tdbCtx   *Context // the tiledb context for all operations
+	arrayURI string   // the uri of the array
+	readOnly bool     // true if opened only for reads
+	closed   bool     // true if the file is closed
+
+	// used when opened for reading
+	arraySize int64 // the size of the array as returned by FileStore.Size
+	bytesRead int64 // the total bytes read so far. Used as an offset for read operations
+
+	// used when opened for writing
+	createSchema bool              // should create array schema on Close
+	mimeType     FileStoreMimeType // the mime type for the created array
+	buf          bytes.Buffer      // internal buffer for data
+
+}
+
+// URI returns the file's uri
+func (f *File) URI() string {
+	return f.arrayURI
+}
+
+// Read satisfies io.Reader
+func (f *File) Read(p []byte) (n int, err error) {
+	if f.closed {
+		return 0, errors.New("file is closed")
+	}
+	if !f.readOnly {
+		return 0, errors.New("file is not open for reads")
+	}
+
+	if len(p) == 0 {
+		return 0, nil
+	}
+
+	bytesRemaining := f.arraySize - f.bytesRead
+	if bytesRemaining == 0 {
+		return 0, io.EOF
+	}
+	if int64(len(p)) > bytesRemaining {
+		p = p[0:bytesRemaining]
+	}
+
+	cArrayURI := C.CString(f.arrayURI)
+	defer C.free(unsafe.Pointer(cArrayURI))
+	if err := bufferExport(f.tdbCtx, cArrayURI, f.bytesRead, p); err != nil {
+		return 0, err
+	}
+	f.bytesRead += int64(len(p))
+
+	return len(p), nil
+}
+
+// Write satisfies io.Writer
+func (f *File) Write(p []byte) (n int, err error) {
+	if f.closed {
+		return 0, errors.New("file is closed")
+	}
+	if f.readOnly {
+		return 0, errors.New("file is not open for writes")
+	}
+	return f.buf.Write(p)
+}
+
+// Close satisfies io.Closer
+func (f *File) Close() error {
+	if f.closed {
+		return nil
+	}
+
+	defer func() {
+		f.closed = true
+	}()
+
+	if f.createSchema {
+		schema, err := NewArraySchemaForFile(f.tdbCtx, "")
+		if err != nil {
+			return err
+		}
+		array, err := NewArray(f.tdbCtx, f.arrayURI)
+		if err != nil {
+			return err
+		}
+		err = array.Create(schema)
+		if err != nil {
+			return err
+		}
+	}
+
+	cArrayURI := C.CString(f.arrayURI)
+	defer C.free(unsafe.Pointer(cArrayURI))
+	return bufferImport(f.tdbCtx, cArrayURI, f.buf.Bytes(), f.mimeType)
+}
+
+// CreateFile creates a TileDB array at arrayURI, suitable to store a file of mimeType, and opens it for writing.
+// The returned File is an inmemory writer. Each Write stores the data in an internal bytes.Buffer.
+// The actual TileDB array creation happens on Close.
+func (fs *FileStore) CreateFile(arrayURI string, mimeType FileStoreMimeType) (*File, error) {
+	return &File{
+		tdbCtx:       fs.tdbCtx,
+		arrayURI:     arrayURI,
+		createSchema: true,
+		mimeType:     mimeType,
+	}, nil
+}
+
+// OpenFileRead opens for reading the array at arrayURI, which should have a filestore schema.
+func (fs *FileStore) OpenFileToRead(arrayURI string) (*File, error) {
```

Comment:
I think I'm still confused on the needs for these functions. We don't really open a file in read/write mode so I don't think we need to distinguish this in any way?